File: packages/analytics/__tests__/analytics-manager.test.ts
import { describe, it, expect, vi } from 'vitest'
import { createAnalytics } from '../src/analytics-manager.js'
import { createMockSink } from '../src/mock-sink.js'
import { createMemoryStorage } from '../src/storage.js'
import { SCHEMA_VERSION } from '../src/types.js'
import { isUuidV4 } from '../src/uuid.js'
import type { AnalyticsConfig } from '../src/types.js'

function devConfig(
  sink = createMockSink(),
  extra: Partial<AnalyticsConfig> = {},
): { a: ReturnType<typeof createAnalytics>; sink: typeof sink } {
  const a = createAnalytics({
    app: 'test-app',
    env: 'development',
    sinks: [sink],
    session: { storage: createMemoryStorage() },
    identity: { storage: createMemoryStorage() },
    consent: { granted: ['analytics'] },
    ...extra,
  })
  return { a, sink }
}

describe('createAnalytics — envelope shape', () => {
  it('emits a canonical Segment track envelope', () => {
    const { a, sink } = devConfig()
    a.track('Signup Clicked', { plan: 'pro' })
    expect(sink.events).toHaveLength(1)
    const ev = sink.events[0]
    expect(ev.type).toBe('track')
    expect(ev.event).toBe('Signup Clicked')
    expect(ev.properties).toMatchObject({ plan: 'pro' })
    expect(isUuidV4(ev.messageId)).toBe(true)
    expect(isUuidV4(ev.anonymousId)).toBe(true)
    expect(isUuidV4(ev.sessionId)).toBe(true)
    expect(ev.schemaVersion).toBe(SCHEMA_VERSION)
    expect(typeof ev.timestamp).toBe('string')
    expect(new Date(ev.timestamp).toString()).not.toBe('Invalid Date')
    expect(ev.context.app).toBe('test-app')
    expect(ev.context.env).toBe('development')
    expect(ev.context.library.name).toBe('@refraction-ui/analytics')
  })

  it('every call type produces the right envelope', () => {
    const { a, sink } = devConfig()
    a.identify('user_1', { plan: 'pro', email: 'a@b.com' })
    a.page('Home', { path: '/' })
    a.screen('Dashboard')
    a.group('org_1', { name: 'Acme' })
    a.alias('user_2', 'user_1')

    const byType = Object.fromEntries(sink.events.map((e) => [e.type, e]))
    expect(byType.identify.traits).toMatchObject({ plan: 'pro' })
    // PII redaction applied to identify traits
    expect(byType.identify.traits!.email).toBe('[REDACTED]')
    expect(byType.identify.userId).toBe('user_1')
    expect(byType.page.event).toBe('Home')
    expect(byType.screen.event).toBe('Dashboard')
    expect(byType.group.groupId).toBe('org_1')
    expect(byType.alias.previousId).toBe('user_1')
    expect(byType.alias.userId).toBe('user_2')
  })

  it('redacts PII in track properties via the deny-list', () => {
    const { a, sink } = devConfig()
    a.track('Form Submitted', { email: 'a@b.com', value: 42 })
    expect(sink.events[0].properties).toEqual({
      email: '[REDACTED]',
      value: 42,
    })
  })

  it('redacts caller-supplied redactKeys', () => {
    const { a, sink } = devConfig(createMockSink(), {
      redactKeys: ['internalScore'],
    })
    a.track('X', { internalScore: 99, ok: 1 })
    expect(sink.events[0].properties).toEqual({
      internalScore: '[REDACTED]',
      ok: 1,
    })
  })
})

describe('createAnalytics — identity & session', () => {
  it('binds userId after identify and clears it on reset', () => {
    const { a } = devConfig()
    expect(a.userId()).toBeUndefined()
    a.identify('user_9')
    expect(a.userId()).toBe('user_9')
    const anonBefore = a.anonymousId()
    a.reset()
    expect(a.userId()).toBeUndefined()
    expect(a.anonymousId()).not.toBe(anonBefore)
  })

  it('exposes a working session API', () => {
    const { a } = devConfig()
    const id1 = a.session.id()
    expect(isUuidV4(id1)).toBe(true)
    const id2 = a.session.start()
    expect(id2).not.toBe(id1)
    a.session.end()
  })

  it('shares one sessionId across events in a session', () => {
    const { a, sink } = devConfig()
    a.track('A')
    a.track('B')
    expect(sink.events[0].sessionId).toBe(sink.events[1].sessionId)
  })
})

describe('createAnalytics — with(context) child', () => {
  it('merges child context into every event without affecting the parent', () => {
    const { a, sink } = devConfig()
    const child = a.with({ feature: 'checkout' } as never)
    child.track('Child Event')
    a.track('Parent Event')
    const childEv = sink.events.find((e) => e.event === 'Child Event')!
    const parentEv = sink.events.find((e) => e.event === 'Parent Event')!
    expect((childEv.context as Record<string, unknown>).feature).toBe(
      'checkout',
    )
    expect((parentEv.context as Record<string, unknown>).feature).toBeUndefined()
  })

  it('child shares identity/session/sinks with the parent', () => {
    const { a, sink } = devConfig()
    const child = a.with({})
    expect(child.anonymousId()).toBe(a.anonymousId())
    child.addSink(createMockSink({ name: 'extra' }))
    expect(a.sinks).toContain('extra')
    expect(sink).toBeDefined()
  })
})

describe('createAnalytics — sink registry', () => {
  it('lists, adds, and removes sinks', () => {
    const { a } = devConfig()
    expect(a.sinks).toContain('mock')
    const extra = createMockSink({ name: 'extra' })
    a.addSink(extra)
    expect(a.sinks).toContain('extra')
    a.removeSink('extra')
    expect(a.sinks).not.toContain('extra')
  })

  it('calls sink.init exactly once before the first deliver', () => {
    const { a, sink } = devConfig()
    a.track('A')
    a.track('B')
    expect(sink.initCalls).toHaveLength(1)
    expect(sink.initCalls[0]).toMatchObject({
      app: 'test-app',
      env: 'development',
    })
  })

  it('auto-registers a built-in http sink when endpoint is set', () => {
    const a = createAnalytics({
      app: 'x',
      env: 'development',
      endpoint: 'https://t.example.com',
      writeKey: 'wk',
      session: { storage: createMemoryStorage() },
      identity: { storage: createMemoryStorage() },
    })
    expect(a.sinks).toContain('http')
  })

  it('dev preset auto-adds a console sink', () => {
    const a = createAnalytics({
      app: 'x',
      env: 'development',
      session: { storage: createMemoryStorage() },
      identity: { storage: createMemoryStorage() },
    })
    expect(a.sinks).toContain('console')
  })
})

describe('createAnalytics — consent gating', () => {
  it('does not deliver to a sink whose categories are not granted', () => {
    const open = createMockSink({ name: 'open' }) // no categories
    const gated = createMockSink({
      name: 'gated',
      consentCategories: ['marketing'],
    })
    const a = createAnalytics({
      app: 'x',
      env: 'development',
      sinks: [open, gated],
      session: { storage: createMemoryStorage() },
      identity: { storage: createMemoryStorage() },
      consent: { granted: [] },
    })
    a.track('A')
    expect(open.events).toHaveLength(1) // no categories → allowed
    expect(gated.events).toHaveLength(0) // marketing not granted → blocked

    a.consent.grant('marketing')
    a.track('B')
    expect(gated.events).toHaveLength(1) // now flowing
  })

  it('per-sink categories are independent', () => {
    const analyticsSink = createMockSink({
      name: 's-analytics',
      consentCategories: ['analytics'],
    })
    const marketingSink = createMockSink({
      name: 's-marketing',
      consentCategories: ['marketing'],
    })
    const a = createAnalytics({
      app: 'x',
      env: 'development',
      sinks: [analyticsSink, marketingSink],
      session: { storage: createMemoryStorage() },
      identity: { storage: createMemoryStorage() },
      consent: { granted: ['analytics'] },
    })
    a.track('A')
    expect(analyticsSink.events).toHaveLength(1)
    expect(marketingSink.events).toHaveLength(0)
  })
})

describe('createAnalytics — batching (prod preset)', () => {
  function prodAnalytics(sink = createMockSink(), batchSize = 3) {
    const a = createAnalytics({
      app: 'x',
      env: 'production',
      preset: 'prod',
      batchSize,
      sinks: [sink],
      session: { storage: createMemoryStorage() },
      identity: { storage: createMemoryStorage() },
      consent: { granted: ['analytics'] },
    })
    return { a, sink }
  }

  it('buffers events and flushes when batchSize is reached', async () => {
    const { a, sink } = prodAnalytics(createMockSink(), 3)
    a.track('1')
    a.track('2')
    expect(sink.events).toHaveLength(0) // still buffered
    a.track('3') // hits batchSize → auto-flush
    await Promise.resolve()
    expect(sink.events).toHaveLength(3)
    expect(sink.deliveries[0].batch).toHaveLength(3)
  })

  it('flush() drains the buffer on demand', async () => {
    const { a, sink } = prodAnalytics(createMockSink(), 100)
    a.track('1')
    a.track('2')
    expect(sink.events).toHaveLength(0)
    await a.flush()
    expect(sink.events).toHaveLength(2)
  })

  it('dev preset delivers synchronously (no batching)', () => {
    const { a, sink } = devConfig()
    a.track('immediate')
    expect(sink.events).toHaveLength(1)
  })
})

describe('createAnalytics — sampling', () => {
  it('drops events below the sample rate', () => {
    const spy = vi.spyOn(Math, 'random').mockReturnValue(0.9)
    const { a, sink } = devConfig(createMockSink(), { sampleRate: 0.5 })
    a.track('dropped')
    expect(sink.events).toHaveLength(0)
    spy.mockReturnValue(0.1)
    a.track('kept')
    expect(sink.events).toHaveLength(1)
    spy.mockRestore()
  })

  it('sampleRate >= 1 keeps everything', () => {
    const { a, sink } = devConfig(createMockSink(), { sampleRate: 1 })
    a.track('a')
    a.track('b')
    expect(sink.events).toHaveLength(2)
  })
})

describe('createAnalytics — noop kill switch', () => {
  it('returns a tree-shakeable noop when enabled:false', () => {
    const sink = createMockSink()
    const a = createAnalytics({
      app: 'x',
      env: 'production',
      enabled: false,
      sinks: [sink],
    })
    expect(a.enabled).toBe(false)
    a.track('nope')
    a.identify('user')
    a.page()
    expect(sink.events).toHaveLength(0)
    expect(a.sinks).toEqual([])
    expect(a.userId()).toBeUndefined()
    // Noop API is fully callable and never throws.
    expect(() => {
      a.session.start()
      a.consent.grant('analytics')
      a.with({}).track('still nothing')
      a.reset()
    }).not.toThrow()
  })

  it('a live collector reports enabled:true', () => {
    const { a } = devConfig()
    expect(a.enabled).toBe(true)
  })
})

File: packages/flutter/test/analytics/analytics_manager_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:refraction_ui/refraction_ui.dart'
    hide createMockSink, MockSinkExtended;
// Test-only mock sink — imported directly (intentionally NOT re-exported by
// the analytics or package barrels, where it would collide with telemetry's
// flat-barrel createMockSink / MockSinkExtended).
import 'package:refraction_ui/src/analytics/mock_sink.dart';

/// Let the dev-preset microtask delivery settle.
Future<void> pump() => Future<void>.delayed(Duration.zero);

({Analytics a, MockSink sink}) devConfig({
  MockSink? sink,
  double sampleRate = 1,
  List<String> redactKeys = const [],
  double Function()? random,
}) {
  final s = sink ?? createMockSink();
  final a = createAnalytics(
    AnalyticsConfig(
      app: 'test-app',
      env: 'development',
      sinks: [s],
      session: SessionConfig(storage: createMemoryStorage()),
      identity: IdentityConfig(storage: createMemoryStorage()),
      consent: const ConsentConfig(granted: ['analytics']),
      sampleRate: sampleRate,
      redactKeys: redactKeys,
      random: random,
    ),
  );
  return (a: a, sink: s);
}

void main() {
  group('createAnalytics — envelope shape', () {
    test('emits a canonical Segment track envelope', () async {
      final c = devConfig();
      c.a.track('Signup Clicked', {'plan': 'pro'});
      await pump();
      expect(c.sink.events, hasLength(1));
      final ev = c.sink.events[0];
      expect(ev.type, AnalyticsEventType.track);
      expect(ev.event, 'Signup Clicked');
      expect(ev.properties, containsPair('plan', 'pro'));
      expect(isUuidV4(ev.messageId), isTrue);
      expect(isUuidV4(ev.anonymousId), isTrue);
      expect(isUuidV4(ev.sessionId), isTrue);
      expect(ev.schemaVersion, kSchemaVersion);
      expect(DateTime.tryParse(ev.timestamp), isNotNull);
      expect(ev.context.app, 'test-app');
      expect(ev.context.env, 'development');
      expect(ev.context.library.name, '@refraction-ui/analytics');
      // Wire JSON shape.
      final json = ev.toJson();
      expect(json['type'], 'track');
      expect(json['schemaVersion'], kSchemaVersion);
      expect((json['context'] as Map)['library'], isNotNull);
    });

    test('every call type produces the right envelope', () async {
      final c = devConfig();
      c.a.identify('user_1', {'plan': 'pro', 'email': 'a@b.com'});
      c.a.page('Home', {'path': '/'});
      c.a.screen('Dashboard');
      c.a.group('org_1', {'name': 'Acme'});
      c.a.alias('user_2', 'user_1');
      await pump();

      final byType = {for (final e in c.sink.events) e.type: e};
      expect(
        byType[AnalyticsEventType.identify]!.traits,
        containsPair('plan', 'pro'),
      );
      // PII redaction applied to identify traits.
      expect(
        byType[AnalyticsEventType.identify]!.traits!['email'],
        '[REDACTED]',
      );
      expect(byType[AnalyticsEventType.identify]!.userId, 'user_1');
      expect(byType[AnalyticsEventType.page]!.event, 'Home');
      expect(byType[AnalyticsEventType.screen]!.event, 'Dashboard');
      expect(byType[AnalyticsEventType.group]!.groupId, 'org_1');
      // group trait `name` is redacted by the exact-match rule.
      expect(byType[AnalyticsEventType.group]!.traits!['name'], '[REDACTED]');
      expect(byType[AnalyticsEventType.alias]!.previousId, 'user_1');
      expect(byType[AnalyticsEventType.alias]!.userId, 'user_2');
    });

    test('redacts PII in track properties via the deny-list', () async {
      final c = devConfig();
      c.a.track('Form Submitted', {'email': 'a@b.com', 'value': 42});
      await pump();
      expect(c.sink.events[0].properties, {'email': '[REDACTED]', 'value': 42});
    });

    test('redacts caller-supplied redactKeys', () async {
      final c = devConfig(redactKeys: ['internalScore']);
      c.a.track('X', {'internalScore': 99, 'ok': 1});
      await pump();
      expect(c.sink.events[0].properties, {
        'internalScore': '[REDACTED]',
        'ok': 1,
      });
    });
  });

  group('createAnalytics — identity & session', () {
    test('binds userId after identify and clears it on reset', () {
      final c = devConfig();
      expect(c.a.userId(), isNull);
      c.a.identify('user_9');
      expect(c.a.userId(), 'user_9');
      final anonBefore = c.a.anonymousId();
      c.a.reset();
      expect(c.a.userId(), isNull);
      expect(c.a.anonymousId(), isNot(anonBefore));
    });

    test('exposes a working session API', () {
      final c = devConfig();
      final id1 = c.a.session.id();
      expect(isUuidV4(id1), isTrue);
      final id2 = c.a.session.start();
      expect(id2, isNot(id1));
      c.a.session.end();
    });

    test('shares one sessionId across events in a session', () async {
      final c = devConfig();
      c.a.track('A');
      c.a.track('B');
      await pump();
      expect(c.sink.events[0].sessionId, c.sink.events[1].sessionId);
    });
  });

  group('createAnalytics — withContext child', () {
    test(
      'merges child context into every event without affecting parent',
      () async {
        final c = devConfig();
        final child = c.a.withContext({'feature': 'checkout'});
        child.track('Child Event');
        c.a.track('Parent Event');
        await pump();
        final childEv = c.sink.events.firstWhere(
          (e) => e.event == 'Child Event',
        );
        final parentEv = c.sink.events.firstWhere(
          (e) => e.event == 'Parent Event',
        );
        expect(childEv.context.extra['feature'], 'checkout');
        expect(parentEv.context.extra['feature'], isNull);
      },
    );

    test('child shares identity/session/sinks with the parent', () {
      final c = devConfig();
      final child = c.a.withContext(const {});
      expect(child.anonymousId(), c.a.anonymousId());
      child.addSink(createMockSink(name: 'extra'));
      expect(c.a.sinks, contains('extra'));
    });
  });

  group('createAnalytics — sink registry', () {
    test('lists, adds, and removes sinks', () {
      final c = devConfig();
      expect(c.a.sinks, contains('mock'));
      c.a.addSink(createMockSink(name: 'extra'));
      expect(c.a.sinks, contains('extra'));
      c.a.removeSink('extra');
      expect(c.a.sinks, isNot(contains('extra')));
    });

    test('calls sink.init exactly once before the first deliver', () async {
      final c = devConfig();
      c.a.track('A');
      c.a.track('B');
      await pump();
      expect(c.sink.initCalls, hasLength(1));
      expect(c.sink.initCalls[0].app, 'test-app');
      expect(c.sink.initCalls[0].env, 'development');
    });

    test('auto-registers a built-in http sink when endpoint is set', () {
      final a = createAnalytics(
        AnalyticsConfig(
          app: 'x',
          env: 'development',
          endpoint: 'https://t.example.com',
          writeKey: 'wk',
          session: SessionConfig(storage: createMemoryStorage()),
          identity: IdentityConfig(storage: createMemoryStorage()),
        ),
      );
      expect(a.sinks, contains('http'));
    });

    test('dev preset auto-adds a console sink', () {
      final a = createAnalytics(
        AnalyticsConfig(
          app: 'x',
          env: 'development',
          session: SessionConfig(storage: createMemoryStorage()),
          identity: IdentityConfig(storage: createMemoryStorage()),
        ),
      );
      expect(a.sinks, contains('console'));
    });
  });

  group('createAnalytics — consent gating', () {
    test(
      'does not deliver to a sink whose categories are not granted',
      () async {
        final open = createMockSink(name: 'open');
        final gated = createMockSink(
          name: 'gated',
          consentCategories: ['marketing'],
        );
        final a = createAnalytics(
          AnalyticsConfig(
            app: 'x',
            env: 'development',
            sinks: [open, gated],
            session: SessionConfig(storage: createMemoryStorage()),
            identity: IdentityConfig(storage: createMemoryStorage()),
            consent: const ConsentConfig(granted: []),
          ),
        );
        a.track('A');
        await pump();
        expect(open.events, hasLength(1));
        expect(gated.events, isEmpty);

        a.consent.grant(['marketing']);
        a.track('B');
        await pump();
        expect(gated.events, hasLength(1));
      },
    );

    test('per-sink categories are independent', () async {
      final analyticsSink = createMockSink(
        name: 's-analytics',
        consentCategories: ['analytics'],
      );
      final marketingSink = createMockSink(
        name: 's-marketing',
        consentCategories: ['marketing'],
      );
      final a = createAnalytics(
        AnalyticsConfig(
          app: 'x',
          env: 'development',
          sinks: [analyticsSink, marketingSink],
          session: SessionConfig(storage: createMemoryStorage()),
          identity: IdentityConfig(storage: createMemoryStorage()),
          consent: const ConsentConfig(granted: ['analytics']),
        ),
      );
      a.track('A');
      await pump();
      expect(analyticsSink.events, hasLength(1));
      expect(marketingSink.events, isEmpty);
    });
  });

  group('createAnalytics — batching (prod preset)', () {
    ({Analytics a, MockSink sink}) prod({int batchSize = 3}) {
      final s = createMockSink();
      final a = createAnalytics(
        AnalyticsConfig(
          app: 'x',
          env: 'production',
          preset: 'prod',
          batchSize: batchSize,
          sinks: [s],
          session: SessionConfig(storage: createMemoryStorage()),
          identity: IdentityConfig(storage: createMemoryStorage()),
          consent: const ConsentConfig(granted: ['analytics']),
        ),
      );
      return (a: a, sink: s);
    }

    test('buffers events and flushes when batchSize is reached', () async {
      final c = prod(batchSize: 3);
      c.a.track('1');
      c.a.track('2');
      expect(c.sink.events, isEmpty); // still buffered
      c.a.track('3'); // hits batchSize → auto-flush
      await pump();
      expect(c.sink.events, hasLength(3));
      expect(c.sink.deliveries[0].batch, hasLength(3));
    });

    test('flush() drains the buffer on demand', () async {
      final c = prod(batchSize: 100);
      c.a.track('1');
      c.a.track('2');
      expect(c.sink.events, isEmpty);
      await c.a.flush();
      expect(c.sink.events, hasLength(2));
    });

    test('dev preset delivers without batching', () async {
      final c = devConfig();
      c.a.track('immediate');
      await pump();
      expect(c.sink.events, hasLength(1));
    });
  });

  group('createAnalytics — sampling', () {
    test('drops events below the sample rate', () async {
      var roll = 0.9;
      final c = devConfig(sampleRate: 0.5, random: () => roll);
      c.a.track('dropped');
      await pump();
      expect(c.sink.events, isEmpty);
      roll = 0.1;
      c.a.track('kept');
      await pump();
      expect(c.sink.events, hasLength(1));
    });

    test('sampleRate >= 1 keeps everything', () async {
      final c = devConfig(sampleRate: 1);
      c.a.track('a');
      c.a.track('b');
      await pump();
      expect(c.sink.events, hasLength(2));
    });
  });

  group('createAnalytics — noop kill switch', () {
    test('returns a noop when enabled:false', () {
      final sink = createMockSink();
      final a = createAnalytics(
        AnalyticsConfig(
          app: 'x',
          env: 'production',
          enabled: false,
          sinks: [sink],
        ),
      );
      expect(a.enabled, isFalse);
      a.track('nope');
      a.identify('user');
      a.page();
      expect(sink.events, isEmpty);
      expect(a.sinks, isEmpty);
      expect(a.userId(), isNull);
      expect(() {
        a.session.start();
        a.consent.grant(['analytics']);
        a.withContext(const {}).track('still nothing');
        a.reset();
      }, returnsNormally);
    });

    test('a live collector reports enabled:true', () {
      expect(devConfig().a.enabled, isTrue);
    });
  });
}
